compiler/noirc_evaluator/src/ssa/function_builder/data_bus.rs

```diff
@@ -44,34 +57,51 @@ impl DataBusBuilder {
     }
 }
 
+#[derive(Clone, Debug)]
+pub(crate) struct CallData {
+    pub(crate) array_id: ValueId,
+    pub(crate) index_map: HashMap<ValueId, usize>,
+}
+
 #[derive(Clone, Default, Debug)]
 pub(crate) struct DataBus {
-    pub(crate) call_data: Option<ValueId>,
-    pub(crate) call_data_map: HashMap<ValueId, usize>,
+    pub(crate) call_data: Vec<CallData>,
     pub(crate) return_data: Option<ValueId>,
 }
 
 impl DataBus {
     /// Updates the databus values with the provided function
     pub(crate) fn map_values(&self, mut f: impl FnMut(ValueId) -> ValueId) -> DataBus {
-        let mut call_data_map = HashMap::default();
-        for (k, v) in self.call_data_map.iter() {
-            call_data_map.insert(f(*k), *v);
-        }
-        DataBus {
-            call_data: self.call_data.map(&mut f),
-            call_data_map,
-            return_data: self.return_data.map(&mut f),
-        }
+        let call_data = self
+            .call_data
+            .iter()
+            .map(|cd| {
+                let mut call_data_map = HashMap::default();
+                for (k, v) in cd.index_map.iter() {
+                    call_data_map.insert(f(*k), *v);
+                }
+                CallData { array_id: f(cd.array_id), index_map: call_data_map }
+            })
+            .collect();
+        DataBus { call_data, return_data: self.return_data.map(&mut f) }
     }
 
+    pub(crate) fn call_data_array(&self) -> Vec<ValueId> {
+        self.call_data.iter().map(|cd| cd.array_id).collect()
+    }
     /// Construct a databus from call_data and return_data data bus builders
-    pub(crate) fn get_data_bus(call_data: DataBusBuilder, return_data: DataBusBuilder) -> DataBus {
-        DataBus {
-            call_data: call_data.databus,
-            call_data_map: call_data.map,
-            return_data: return_data.databus,
+    pub(crate) fn get_data_bus(
+        call_data: Vec<DataBusBuilder>,
+        return_data: DataBusBuilder,
+    ) -> DataBus {
+        let mut call_data_args = Vec::new();
+        for call_data in call_data {
```

Comment:
```suggestion
        for call_data_item in call_data {
```
^ using a distinct name for the items could make this easier to read